sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveTableScanExec.scala

```diff
@@ -0,0 +1,501 @@
+      maxReaderBatchSizeBytes = rapidsConf.maxReadBatchSizeBytes,
+      metrics = allMetrics,
+      options = options
+    )
+
+    PartitionReaderIterator.buildReader(readerFactory)
+  }
+
+  /**
+   * Prune output projection to those columns that are to be read from
+   * the input file/buffer, in the same order as [[requestedAttributes]]
+   * Removes partition columns, and returns the resultant schema
+   * as a [[StructType]].
+   */
+  private def getRequestedOutputDataSchema(tableSchema: StructType,
+                                           partitionAttributes: Seq[Attribute],
+                                           requestedAttributes: Seq[Attribute]): StructType = {
+    // Read schema in the same order as requestedAttributes.
+    // Note: This might differ from the column order in `tableSchema`.
+    //       In fact, HiveTableScanExec specifies `requestedAttributes` alphabetically.
+    val partitionKeys: HashSet[String] = HashSet() ++ partitionAttributes.map(_.name)
+    val requestedCols = requestedAttributes.filter(a => !partitionKeys.contains(a.name))
+                                           .toList
+    val distinctColumns = requestedCols.distinct
+    val distinctFields  = distinctColumns.map(a => tableSchema.apply(a.name))
+    StructType(distinctFields)
+  }
+
+  private def createReadRDDForDirectories(readFile: PartitionedFile => Iterator[InternalRow],
+                                          directories: Seq[(URI, InternalRow)],
+                                          readSchema: StructType,
+                                          sparkSession: SparkSession,
+                                          hadoopConf: Configuration): RDD[InternalRow] = {
+    def isNonEmptyDataFile(f: FileStatus): Boolean = {
+      if (!f.isFile || f.getLen == 0) {
+        false
+      }
+      else {
+        val name = f.getPath.getName
+        !((name.startsWith("_") && !name.contains("=")) || name.startsWith("."))
+      }
+    }
+
+    val filePartitions: Seq[FilePartition] = directories.flatMap { case (directory, partValues) =>
+      val path               = new Path(directory)
+      val fs                 = path.getFileSystem(hadoopConf)
+      val dirContents        = fs.listStatus(path).filter(isNonEmptyDataFile)
+      val partitionDirectory = PartitionDirectory(partValues, dirContents)
+      val maxSplitBytes      = FilePartition.maxSplitBytes(sparkSession, Array(partitionDirectory))
+
+      val splitFiles: Seq[PartitionedFile] = dirContents.flatMap { f =>
+        PartitionedFileUtil.splitFiles(
+          sparkSession,
+          f,
+          f.getPath,
+          isSplitable = true,
+          maxSplitBytes,
+          partitionDirectory.values
+        )
+      }.sortBy(_.length)(implicitly[Ordering[Long]].reverse)
+      FilePartition.getFilePartitions(sparkSession, splitFiles, maxSplitBytes)
+    }
+
+    // TODO [future]: Handle small-file optimization.
```

Comment:
Do we need a follow on issue for this?

Reply:
Raised this issue: https://github.com/NVIDIA/spark-rapids/issues/7017